### ISSUE TITLE
Bug Report: Reshard command stuck in copy phase for very large table

### ISSUE THREAD

**Issue by brirams** [vitessio/vitess#10012]
#### Overview 

We're resharding one of our largest keyspaces by dataset size and are seeing the copy phase spin in
a retry loop due to transport errors at the destination primaries:
```
E0324 09:21:05.267377    3958 vreplicator.go:144] Replicate error: vttablet: rpc error: code = Unavailable desc = transport is closing
```

We suspect that this is due to the streaming query that ends up running at the source timing out
because of:
- the compound PK
- the distribution of data across those columns
- each row containing 3 big columns of types `mediumtext`, `longblob` and `blob`

#### Context
- issue discussed in vitess slack: https://vitess.slack.com/archives/CMKTCUYNQ/p1647022135431159
- this single shard has a single table that is 2.4TB in size

- number of rows:
```
mysql> select count(1) from table;
+------------+
| count(1)   |
+------------+
| 2662419503 |
+------------+
1 row in set (32 min 0.31 sec)
```

- schema:
```
CREATE TABLE `table` (
  `id1` bigint(20) unsigned NOT NULL,
  `id2` bigint(20) unsigned NOT NULL,
  `id3` bigint(20) unsigned NOT NULL,
  `col1` bigint(20) unsigned NOT NULL,
  `col2` bigint(20) unsigned NOT NULL,
  `col3` int(10) unsigned NOT NULL,
  `col4` varchar(255) NOT NULL,
  `col5` tinyint(3) unsigned NOT NULL,
  `col6` bigint(20) unsigned NOT NULL,
  `col7` bigint(20) unsigned NOT NULL,
  `col8` bigint(20) unsigned NOT NULL,
  `col9` bigint(20) unsigned NOT NULL,
  `col10` bigint(20) unsigned NOT NULL,
  `col11` bigint(20) unsigned NOT NULL,
  `bigcol1` mediumtext NOT NULL,
  `bigcol2` longblob,
  `bigcol3` blob,
  `col12` varchar(255) NOT NULL,
  `col13` tinyint(3) unsigned NOT NULL,
  `col14` tinyint(3) unsigned NOT NULL,
  PRIMARY KEY (`id1`,`id2`,`id3`),
  KEY `idx1` (`coll11`),
  KEY `idx2` (`id2`,`id3`),
  KEY `idx3` (`col10`),
  KEY `idx4` (`id1`,`id2`,`col5`,`id3`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 ROW_FORMAT=COMPRESSED
```

- distribution of data: each `id1` has on the order of dozens to hundreds and rarely thousands of
`id2`s. each combo of `id1/id2` has on the order of hundreds to tens of thousands of instances of `id3`
```
mysql> select count(distinct id1) from vt_$keyspace.$table;
+-------------------------+
| count(distinct id1)     |
+-------------------------+
|                 1626859 |
+-------------------------+
1 row in set (32.64 sec)

mysql> select count(distinct id2) from vt_$keyspace.$table;
+----------------------------+
| count(distinct id2)        |
+----------------------------+
|                    7994720 |
+----------------------------+
1 row in set (3 min 33.96 sec)

mysql> select count(distinct timestamp) from vt_$keyspace.$table;
+---------------------------+
| count(distinct timestamp) |
+---------------------------+
|                2550627341 |
+---------------------------+
1 row in set (1 hour 21 min 50.15 sec)
```

#### Things we've tried

- bumping the `max_allowed_packet` at the source to 1GB
- bumping `net_read_timeout` and `net_write_timeout` at the source to 24hr
- we built an experimental patch of our version of vitess and added a limit/offset to the query that
  the `rowstreamer` executes. this allowed 7/8 streams to finish but the last one eventually hit the
  same issue

### Reproduction Steps

- spin up destination tablets in 1 primary + 3 replica configuration
- we're resharding shard `94-95` 8 ways:
```
sudo vtctlclient \
    -grpc_auth_static_client_creds "creds.json" \
    -server localhost:$port \
    Reshard -cells $cell -tablet_types rdonly,replica keyspace.reshard_94-95 94-95 9400-9420,9420-9440,9440-9460,9460-9480,9480-94a0,94a0-94c0,94c0-94e0,94e0-9500
```
- vreplication rows are created in in destination tablets and vcopier starts to kick off until all
  destination primaries starts throwing `transport is closing` errors and gets stuck in a retry loop with the error:
```
E0324 09:21:05.271556    3958 controller.go:162] stream 1: Code: UNAVAILABLE
vttablet: rpc error: code = Unavailable desc = transport is closing
, retrying after 5s
```

### Binary Version

```sh
Version 10.0.2 from our fork: https://github.com/tinyspeck/vitess/commit/80454805d95cb12999e27ea66ae3069f5900c5b6

$ vttablet --version
Version: 10.0.2 (Jenkins build $N) (Git revision 80454805d branch 'HEAD') built on Fri Jul 23 09:23:38 UTC 2021 by $user using go1.15.6 linux/amd64
```

### Operating System and Environment details

```sh
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.6 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.6 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

$  uname -sr
Linux 5.4.0-1068-aws

$ uname -m
x86_64
```

### Log Fragments

- from destination primary for shard `94e0-9500`: https://gist.github.com/brirams/e5dcfc4fef96505ec93d134f9b7a96dd

- from the source rdonly for shard `94-95`: https://gist.github.com/brirams/7081c0b2068b664576c3692ee1c6cf63

- state of destination primary vrep rows:
```
mysql> select * from vreplication\G;
*************************** 1. row ***************************
                   id: 1
             workflow: reshard_94-95
               source: keyspace:"$keyspace" shard:"94-95" filter:<rules:<match:"/.*" filter:"94e0-9500" > >
                  pos: MySQL56/019b408b-a1ee-11eb-84b0-06e74ae44741:1-118470730,2dbead88-d511-11e9-b38e-0a3775e70552:1-6982850,3ebceff9-54a8-11ec-a871-12e77809e3bd:1-650162,45dfea66-0b5f-11ea-bd64-12552500829f:1-549292666,46c5df4f-251a-11ec-b4bf-0e5a0bca87f3:1-65771443,4f2c167e-38fe-11ec-a6ee-1203449837c9:1-696656,54168235-23cd-11ec-8d02-125498d4cbf1:1-43968699,5b6ac31b-23ec-11ec-a520-0a018519c54b:1-56160064,659b541b-a208-11eb-aabb-125c8fefeba3:1-295082947,68578cf4-a191-11eb-869e-0e3b8625fa63:1-41711662,6df68cfa-d63a-11e9-9c0f-0e197304d3d0:1-55018692,73ed64dd-0bc3-11ea-8090-0e6dd52c0433:1-28202936,7ac6c124-2e26-11ea-aac3-0adf91de2905:1-268213319,9570605c-df11-11e9-a836-12904c199154:1-35565225,9f626a0b-5482-11ec-8c68-0e25275c100d:1-54723190,ac777cd8-271a-11ec-a625-068bfd53a9e9:1-28331302,ba5f189f-5d49-11ec-acb8-0669b657b6d9:1-230823840,bd026bbf-aa87-11ec-823c-0eaefe5704db:1-5216700,c83bceda-966c-11ec-b350-0ea68e1a9cf9:1-1392726,eb1fb564-0b5e-11ea-b48f-06d5644ad83b:1-320833702,fe70246d-d510-11e9-be11-06579990e326:1-11746022
             stop_pos: NULL
              max_tps: 9223372036854775807
  max_replication_lag: 9223372036854775807
                 cell: $cell
         tablet_types: rdonly,replica
         time_updated: 1648138732
transaction_timestamp: 1648138732
                state: Copying
              message:
              db_name: vt_$keyspace
1 row in set (0.00 sec)

ERROR:
No query specified

mysql> select * from copy_state\G;
*************************** 1. row ***************************
  vrepl_id: 1
table_name: $table
    lastpk: fields:<name:"id1" type:UINT64 > fields:<name:"id2" type:UINT64 > fields:<name:"id3" type:UINT64 > rows:<lengths:10 lengths:12 lengths:16 values:"21896866681713164620811493120920874010" >
1 row in set (0.01 sec)
```

- mysql explain at the source:
```
mysql> explain select <all-columns> from $table where (id1 = 2391670107 and id2 = 2694501656150 and id3 > 1636418777001000) or (id1 = 2391670107 and id2 > 2694501656150) or (id1 > 2391670107) order by id1,id2,id3;
+----+-------------+----------+------------+-------+----------------------------------+---------+---------+------+------------+----------+-------------+
| id | select_type | table    | partitions | type  | possible_keys                    | key     | key_len | ref  | rows       | filtered | Extra       |
+----+-------------+----------+------------+-------+----------------------------------+---------+---------+------+------------+----------+-------------+
|  1 | SIMPLE      | $table   | NULL       | range | PRIMARY,$idx2,$idx4              | PRIMARY | 24      | NULL | 1815103128 |   100.00 | Using where |
+----+-------------+----------+------------+-------+----------------------------------+---------+---------+------+------------+----------+-------------+
1 row in set, 1 warning (0.02 sec)
```

**Comment by makinje16** [comment]
Dropping an update on everything we have tried so far. 

* Add vreplication_rowstreamer_limit
    * In https://github.com/tinyspeck/vitess/pull/240 we introduced a new flag, mentioned above, to include a limit to the query created in the `buildSelect` function found in `rowstreamer.go`. As of right now we have that limit tuned to 4000000 rows which does allow queries to complete however after a given amount of time no more data is being copied.
* Increase max allowed packet to 1GB
    * This was recommended to us since the table being copied has blob, longblob, and mediumText columns. There were no indications of packet size being too small. In order to double check this and confirm that the queries we are stuck on were not exceeding this value, we looked at one query found from show full processlist and found that the largest row length was 202 bytes much lower than the 1GB that was set.
* Net Read/Write to 24hrs
    * A common practice for us is to increase net read and write timeouts to 1hr since we have had issues with our default of 30seconds. Due to the nature of this reshard and some of the queries taking over 6hrs before the limit was introduced we bumped both of these values to 24hrs to no avail.
* grpc_keepalive_timeout to 600s
    * We increased this value since in [slack](https://vitess.slack.com/archives/CMKTCUYNQ/p1648757009241979?thread_ts=1647022135.431159&cid=CMKTCUYNQ) the `grpc_server_keepalive_enforcement_policy_min_time` (30s) and `grpc_keepalive_timeout`, defaulted to 10s, may be in confluence with each other leading to 40s timeouts which correlated with what we saw in the amount of time elapsing before a queries connection was being aborted.


In the most recent state we are now copying, one stream completes while the other 
7/8 seem to flatline in the number of rows copied, using vttablet_v_replication_copy_row_count as the reference. Copying always starts off seemingly well and fast but then halts after some time on queries that are returning in under 40s when ran from a mysql prompt on the source tablet. We think the issues are stemming from one of two places now since with the limit, queries seem to be returning in a reasonable amount of time. The first is in grpc land in VStreamRows in which source rows are being sent to the destination, or, the events/rows are not being applied once received. I lean towards grpc since there have been no indicators in the logs regarding binlog events not being applied.

**Comment by maksimov** [comment]
So after a rather long investigation we have found that in our case of quite a large table (2.7B rows) and a composite PK (which lead to a full table scan during source streaming), the destinations never had a chance to start writing in the first hour defined by `-vreplication_copy_phase_duration` so the lastPK has never changed, thus each copy phase had to start from the top again.

After some trial and error we have found that the writes begin after the first 3 hours of streaming (over 1B events as reported by rowstreamer), so we raised `-vreplication_copy_phase_duration` (8h) and the MySQL wait_timeout on the destinations (24h), because vstreamer is using the same MySQL connection from the start of vrep throughout.

The `transport is closing` errors were to do with the grpc keep-alive setup, which we have fixed by setting these on src&dest tablets:
```
-grpc_server_keepalive_enforcement_policy_min_time 2m \
-grpc_keepalive_time 10m \
-grpc_keepalive_timeout 10m \
```
Thanks @mattlord for all your help with this!
